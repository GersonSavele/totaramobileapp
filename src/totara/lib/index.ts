/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import config from "./config";
import { Logger, Log } from "./logger";
import { showMessage } from "./tools";
import * as constants from "./constants";

export { config, Logger, Log, showMessage, constants };
