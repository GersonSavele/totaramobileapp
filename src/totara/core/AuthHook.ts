/*
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Jun Yamog <jun.yamog@totaralearning.com>
 */
import { useEffect, useReducer, useRef } from "react";
import { ApolloClient } from "apollo-client";
import { gql } from "apollo-boost";
import { NormalizedCacheObject } from "apollo-cache-inmemory";
import SplashScreen from "react-native-splash-screen";

import { config, Log } from "@totara/lib";
import { AppState, SiteInfo } from "@totara/types";

/**
 * Custom react hook and its primary role is managing the auth state which is in authContextState.
 * Provides state lift functions to change state
 *
 * Requires external dep routines
 *
 * @param bootstrap
 * @param registerDevice
 * @param deviceCleanup
 * @param createApolloClient
 *
 * @returns {
 *  authContextState - key state of auth
 *  onLoginSuccess - on a successful login, pass setup parameter and it would do device registration process
 *  onLoginFailure - call this function if an error can't be handled locally
 *  logOut - logout the current user and start clean up process locally and on the server
 *  apolloClient - an authenticated apollo client
 * }
 */
export const useAuth = (
  bootstrap: () => Promise<AppState | undefined>,
  registerDevice: (setup: Setup) => Promise<AppState>,
  deviceCleanup: (deviceDelete: () => Promise<any>) => Promise<boolean>,
  createApolloClient: (
    apiKey: string,
    uri: string,
    logOut: (localOnly: boolean) => Promise<void>
  ) => ApolloClient<NormalizedCacheObject>
) => ({ initialState }: Props) => {

  const [authContextState, dispatch] = useReducer(
    authContextReducer,
    initialState
  );

  // instance of apollo client this is valid as per faq
  // https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables
  const apolloClient = useRef<ApolloClient<NormalizedCacheObject> | null>(null);

  /**
   * call logOut to clean any state of auth
   */
  const logOut = async (localOnly: boolean = false) => {
    // TODO MOB-231 should remove this localOnly flag, a bit of a hack
    Log.debug("logging out");
    const mutationPromise = () =>
      apolloClient.current && !localOnly
        ? apolloClient.current.mutate({
            mutation: deleteDevice
          })
        : Promise.resolve({ data: { delete_device: true } });

    return deviceCleanup(mutationPromise).then(() =>
      dispatch({ type: "deRegister" })
    );
  };

  // create or stop apolloClient depending on the state and if existing apolloClient has been appState
  if (
    authContextState.appState &&
    authContextState.appState.apiKey &&
    authContextState.isAuthenticated &&
    apolloClient.current === null
  ) {
    Log.debug("creating apolloClient");
    apolloClient.current = createApolloClient(
      authContextState.appState.apiKey,
      authContextState.appState.host,
      logOut
    );
  } else if (apolloClient.current && !authContextState.isAuthenticated) {
    Log.debug("stopping apolloClient");
    // TODO MOB-231 review this, possibly better implementation
    apolloClient.current.stop();
    apolloClient.current = null;
  }

  /**
   * When a successful login is achieved call this with setup it would start the login sequence and right authenticated state
   *
   * Can be used on either a AuthComponent react component or a promise chain
   *
   * @param setup
   */
  const onLoginSuccess = async (setup: Setup) => {
    if (authContextState.isAuthenticated) {
      await logOut();
    }

    dispatch({ type: "register", payload: setup });
  };

  /**
   * When a failure of login happens call this
   *
   * @param error
   */
  const onLoginFailure = async (error: Error) => {
    Log.error("login failure, bringing back to initial loading state", error);

    dispatch({ type: "authError" });
  };

  /**
   * turn off the splash screen and bootstrap the provider
   */
  useEffect(() => {
    const doBootStrap = () => {
      Log.debug("doBootStrap", authContextState);
      if (SplashScreen) SplashScreen.hide();
      bootstrap().then(appState => {
        dispatch({ type: "bootstrap", payload: appState });
      });
    };

    if (authContextState.authStep === "loading") doBootStrap();
  }, [authContextState.authStep]);

  /**
   * When setup has been initialiazed, perform a side affect to get the apikey
   */
  useEffect(() => {
    const doRegisterDevice = () => {
      Log.debug("doRegisterDevice", authContextState);
      if (authContextState.setup) {
        registerDevice(authContextState.setup).then(appState => {
          dispatch({ type: "registered", payload: appState });
        }).catch(error => {
            Log.error("error on registering", error);
            dispatch({ type: "authError" });
          }
        )
      }
    };

    if (authContextState.authStep === "setupSecretInit")
      doRegisterDevice();
  }, [authContextState.authStep, authContextState.isAuthenticated]);

  return {
    authContextState,
    onLoginSuccess,
    onLoginFailure,
    logOut,
    apolloClient: apolloClient.current
  };
};

const authContextReducer = (state: AuthContextState, action: Action): AuthContextState => {
  Log.debug("authContextReducer: state", state, "action", action);
  switch (action.type) {
    case "register": {
      if (action.payload && "secret" in action.payload)
        return {
          ...state,
          setup: action.payload,
          authStep: "setupSecretInit"
        };
      else throw new Error(`unexpected payload in action ${action}`);
    }

    case "registered": {
      if (action.payload && "apiKey" in action.payload)
        return {
          ...state,
          appState: action.payload,
          isAuthenticated: true,
          authStep: "setupDone"
        };
      else throw new Error(`unexpected payload in action ${action}`);
    }

    case "bootstrap": {
      if (action.payload && "apiKey" in action.payload)
        return {
          ...state,
          appState: action.payload,
          isAuthenticated: true,
          isLoading: false,
          authStep: "bootstrapDone"
        };
      else
        return {
          ...state,
          isAuthenticated: false,
          isLoading: false,
          authStep: "bootstrapDone"
        };
    }

    case "deRegister":
      return {
        ...state,
        appState: undefined,
        isAuthenticated: false,
        isLoading: false,
        authStep: "bootstrapDone"
      };

    case "authError":
      return {
        ...state,
        setup: undefined,
        appState: undefined,
        isAuthenticated: false,
        isLoading: false,
        authStep: "authError"
      };

    case "reload":
      return initialState;
  }
};

type Action = {
  type: "register" | "registered" | "bootstrap" | "deRegister" | "reload" | "authError";
  payload?: AppState | Setup;
};

export const deleteDevice = gql`
  mutation totara_mobile_delete_device {
    delete_device: totara_mobile_delete_device
  }
`;

type Props = {
  initialState: AuthContextState;
};

export type AuthContextState = {
  isLoading: boolean;
  isAuthenticated: boolean;
  appState?: AppState;
  setup?: Setup;
  authStep:
    | "loading"
    | "bootstrapDone"
    | "setupSecretInit"
    | "setupDone"
    | "authError";
};

export const initialState: AuthContextState = {
  appState: undefined,
  setup: undefined,
  isLoading: true,
  isAuthenticated: false,
  authStep: "loading"
};

export interface Setup {
  secret: string;
  uri: string;
  siteInfo: SiteInfo;
}
