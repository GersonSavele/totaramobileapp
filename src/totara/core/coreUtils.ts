/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2021 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { config } from "@totara/lib";

enum Compatible {
  Api = 1
}

const isValidApiVersion = (apiVersoin?: string) => {
  const compatibilityList = isCompatible(apiVersoin);
  return compatibilityList.length > 0;
};

const isCompatible = (version?: string) => {
  const fullCompatible = [Compatible.Api];
  if (config.minApiVersion === "disabled") {
    return fullCompatible;
  } else {
    if (version && config.minApiVersion <= version) return [Compatible.Api];
    else return [];
  }
};

export { isValidApiVersion, isCompatible }