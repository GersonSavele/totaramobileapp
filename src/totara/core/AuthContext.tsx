/*
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Jun Yamog <jun.yamog@totaralearning.com>
 */


import React from "react";

import { config } from "@totara/lib";

import { AuthContextState, initialState, Setup } from "./AuthHook";

export const AuthContext = React.createContext<State>({
  authContextState: initialState,
  logOut: () => {
    return Promise.resolve();
  },
  onLoginSuccess: () => {},
  onLoginFailure: () => {},
});

type State = {
  authContextState: AuthContextState;
  logOut: (local?: boolean) => Promise<void>;
  onLoginSuccess: (setup: Setup) => void;
  onLoginFailure: (error: Error) => void;
}

enum Compatible {
  Api = 1
}

export const AuthConsumer = AuthContext.Consumer;

//TODO MOB-273 Need to integrate correct logic
export const isValidApiVersion = (apiVersoin?: string) => {
  const compatibilityList = isCompatible(apiVersoin);
  return compatibilityList.length > 0;
};

export const isCompatible = (version?: string) => {
  const fullCompatible = [Compatible.Api];
  if (config.minApiVersion === "disabled") {
    return fullCompatible;
  } else {
    if (version && config.minApiVersion <= version) 
      return [Compatible.Api];
    else 
    return [];
  }
};