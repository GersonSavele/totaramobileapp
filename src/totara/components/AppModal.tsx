/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author: Kamala Tennakoon <kamala.tennakoon@totaralearning.com>
 */

import React, { useContext, useState } from "react";

import { translate } from "@totara/locale";
import { AuthContext, Setup } from "@totara/auth/AuthContext";
import { InfoModal} from "./infoModal"
import PrimaryButton from "./PrimaryButton";
import { config } from "@totara/lib";

enum Compatible {
  Api = 1
}

const AppModal = () => {
  const {setup} = useContext(AuthContext);
  const isShowIncompatibleApi = !isValidApiVersion(setup);
  const [isVisible, setIsVisible] = useState(isShowIncompatibleApi);
  if(isVisible) 
    return (
    <InfoModal title={translate("message.sorry")} description={translate("message.incompatible_api")} imageType={"url_not_valid"} visible={isVisible}>
      <PrimaryButton  text={translate("general.try_again")} onPress={() => { setIsVisible(!isVisible) }} />
    </InfoModal>
    );
  else 
    return null;
};

//TODO-Need to integrate correct logic
export const isValidApiVersion = (setup?: Setup) => {
  if (setup && setup.apiVersion) {
    const compatibilityList = isCompatible(setup.apiVersion)
    return compatibilityList.length > 0
  }
  return false;
};

export const isCompatible = (version: string) => {
  if (config.minApiVersion <= version)
    return [Compatible.Api]
  else 
    return []
};

export default AppModal;