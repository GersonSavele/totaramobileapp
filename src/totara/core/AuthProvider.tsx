/*
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Jun Yamog <jun.yamog@totaralearning.com>
 */


import React, { ReactNode } from "react";
import { AsyncStorageStatic } from "@react-native-community/async-storage";

import { AuthContext } from "./AuthContext";
import { useAuth, initialState } from "./AuthHook";
import { registerDevice, deviceCleanup, bootstrap, createApolloClient, fetchData } from "./AuthRoutines";


/**
 * AuthProvider will provide the auth states and data needed for any components inside
 * the scope.  Auth states, methods should be use with AuthConsumer
 *
 * This will also provide the switching from a component used to authenticate a user,
 * Once authenticated it would load the children with an ApolloProvider with a authenticated
 * client.
 *
 * @example
 *
 * <AuthProvider>
 *   <ElementsThatWillHaveAuthenticatedUserAndGraphQLApi/>
 * </AuthProvider>
 *
 */
export const AuthProvider = ( {asyncStorage, children}: Props) => {

  const {
    authContextState,
    onLoginSuccess,
    onLoginFailure,
    logOut,
    apolloClient
    // get fetch from global namespace
    // eslint-disable-next-line no-undef
  } = useAuth(bootstrap(asyncStorage), registerDevice(fetchData(fetch), asyncStorage), deviceCleanup(asyncStorage), createApolloClient)({initialState: initialState });

  const providerValue = {
    isAuthenticated: authContextState.isAuthenticated,
    appState: authContextState.appState,
    authContextState: authContextState,
    logOut: logOut,
    onLoginSuccess: onLoginSuccess,
    onLoginFailure: onLoginFailure,
    apolloClient: apolloClient
  };

  return (
    <AuthContext.Provider value={providerValue}>
      {children}
    </AuthContext.Provider>
  );
};


type Props = {
  children: ReactNode;
  asyncStorage: AsyncStorageStatic;
};
