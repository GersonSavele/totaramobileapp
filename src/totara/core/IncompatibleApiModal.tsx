/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import React from "react";

import { translate } from "@totara/locale";
import { useSession } from "@totara/core";
import { isValidApiVersion } from "@totara/core/coreUtils";
import { InfoModal, PrimaryButton, TertiaryButton } from "@totara/components";
import { config } from "@totara/lib";
import { Linking, ImageSourcePropType } from "react-native";
import { Images } from "@resources/images";

type Props = {
  onCancel?: () => void;
  siteUrl?: string;
  testID?: string;
};

const IncompatibleApiModal = ({ onCancel, siteUrl, testID }: Props) => {
  const { siteInfo, host } = useSession();
  const isShowIncompatibleApi = siteInfo ? !isValidApiVersion(siteInfo.version) : true;
  const site = host || siteUrl;
  if (isShowIncompatibleApi)
    return (
      <InfoModal
        title={translate("incompatible_api.title")}
        description={translate("incompatible_api.description")}
        imageSource={Images.urlNotValid as ImageSourcePropType}
        visible={isShowIncompatibleApi}
        testID={testID}>
        {site && (
          <PrimaryButton
            text={translate("incompatible_api.action_primary")}
            onPress={() => {
              Linking.openURL(config.loginUri(site));
            }}
          />
        )}
        <TertiaryButton
          text={translate("incompatible_api.action_tertiary_cancel")}
          onPress={() => {
            onCancel && onCancel();
          }}
        />
      </InfoModal>
    );
  else return null;
};

export default IncompatibleApiModal;
