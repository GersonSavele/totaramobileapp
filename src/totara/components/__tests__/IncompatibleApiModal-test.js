/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import React from "react";
import renderer from "react-test-renderer";

import AppModal from "../IncompatibleApiModal";
import { AuthContext } from "@totara/core";
import { config } from "@totara/lib";

describe("IncompatibleApiModal", () => {
  const authContextState = {
    isAuthenticated: false
  };

  it("it should not show any modal screens for valid apiVersion or disabled minApiVersion", async () => {
    config.minApiVersion = "2019101802";
    const validSetupVersion = { apiVersion: "2019101802" };
    const validVersionComponent = renderer.create(
      <AuthContext.Provider
        value={{
          setup: validSetupVersion,
          authContextState: authContextState
        }}>
        <AppModal />
      </AuthContext.Provider>
    );
    expect(validVersionComponent.toJSON()).toMatchSnapshot();

    const higherSetupVersion = { apiVersion: "2030101802" };
    const higherVersionComponent = renderer.create(
      <AuthContext.Provider
        value={{
          setup: higherSetupVersion,
          authContextState: authContextState
        }}>
        <AppModal />
      </AuthContext.Provider>
    );
    expect(higherVersionComponent.toJSON()).toMatchSnapshot();

    config.minApiVersion = "disabled";
    const disabledAppMinVersionComponent = renderer.create(
      <AuthContext.Provider
        value={{
          setup: validSetupVersion,
          authContextState: authContextState
        }}>
        <AppModal />
      </AuthContext.Provider>
    );
    expect(disabledAppMinVersionComponent.toJSON()).toMatchSnapshot();
  });
  it("it should show error modal screens for incompatible version", async () => {
    config.minApiVersion = "2019101802";
    const oldSetupVersion = { apiVersion: "2010101802" };
    const oldVersionComponent = renderer.create(
      <AuthContext.Provider value={{ setup: oldSetupVersion, authContextState: authContextState }}>
        <AppModal />
      </AuthContext.Provider>
    );
    expect(oldVersionComponent.toJSON()).toMatchSnapshot();
  });
});
