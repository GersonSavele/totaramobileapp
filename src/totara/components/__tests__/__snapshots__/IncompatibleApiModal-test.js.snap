// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncompatibleApiModal it should not show any modal screens for valid apiVersion or disabled minApiVersion 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.7)",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 4,
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
              "marginHorizontal": 20,
              "marginLeft": "5%",
              "marginRight": "5%",
              "marginVertical": 32,
            },
            Object {
              "backgroundColor": "#FFFFFF",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/resources/images/url_not_valid/url_not_valid.png",
              }
            }
            style={
              Object {
                "alignItems": "center",
                "height": 375,
                "resizeMode": "contain",
                "width": 525,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 34,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              translated incompatible_api.title with undefined
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 17,
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              translated incompatible_api.description with undefined
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "marginVertical": 16,
              "minHeight": 104,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                this is a mock button
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`IncompatibleApiModal it should not show any modal screens for valid apiVersion or disabled minApiVersion 2`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.7)",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 4,
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
              "marginHorizontal": 20,
              "marginLeft": "5%",
              "marginRight": "5%",
              "marginVertical": 32,
            },
            Object {
              "backgroundColor": "#FFFFFF",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/resources/images/url_not_valid/url_not_valid.png",
              }
            }
            style={
              Object {
                "alignItems": "center",
                "height": 375,
                "resizeMode": "contain",
                "width": 525,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 34,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              translated incompatible_api.title with undefined
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 17,
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              translated incompatible_api.description with undefined
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "marginVertical": 16,
              "minHeight": 104,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                this is a mock button
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`IncompatibleApiModal it should not show any modal screens for valid apiVersion or disabled minApiVersion 3`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.7)",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 4,
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
              "marginHorizontal": 20,
              "marginLeft": "5%",
              "marginRight": "5%",
              "marginVertical": 32,
            },
            Object {
              "backgroundColor": "#FFFFFF",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/resources/images/url_not_valid/url_not_valid.png",
              }
            }
            style={
              Object {
                "alignItems": "center",
                "height": 375,
                "resizeMode": "contain",
                "width": 525,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 34,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              translated incompatible_api.title with undefined
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 17,
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              translated incompatible_api.description with undefined
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "marginVertical": 16,
              "minHeight": 104,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                this is a mock button
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`IncompatibleApiModal it should show error modal screens for incompatible version 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.7)",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 4,
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
              "marginHorizontal": 20,
              "marginLeft": "5%",
              "marginRight": "5%",
              "marginVertical": 32,
            },
            Object {
              "backgroundColor": "#FFFFFF",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/resources/images/url_not_valid/url_not_valid.png",
              }
            }
            style={
              Object {
                "alignItems": "center",
                "height": 375,
                "resizeMode": "contain",
                "width": 525,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 32,
              "marginVertical": 16,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 34,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              translated incompatible_api.title with undefined
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#3D444B",
                    "fontSize": 17,
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              translated incompatible_api.description with undefined
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "marginVertical": 16,
              "minHeight": 104,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                this is a mock button
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
